lib/src/internal/http/http_bucket.dart

```diff
@@ -9,86 +9,67 @@ import 'package:nyxx/src/internal/http/http_handler.dart';
+    String? bucketId = getBucketIdFromHeaders(response.headers);
+    return limit == null || remaining == null || reset == null || resetAfter == null || bucketId == null ? null : HttpBucket(limit, remaining, reset, resetAfter, bucketId);
+  }
 
-    // Execute request
-    try {
-      final response = await _httpHandler.httpClient.send(await request.prepareRequest());
+  static String? getBucketIdFromHeaders(Map<String, String> headers) => headers[xRateLimitBucket];
 
-      _setBucketValues(response.headers);
-      return response;
-    } on HttpClientException catch (e) {
-      if (e.response == null) {
-        _httpHandler.logger.warning("Http Error on endpoint: ${request.uri}. Error: [${e.message.toString()}].");
-        return Future.error(e);
-      }
+  static int? getLimitFromHeaders(Map<String, String> headers) => headers[xRateLimitLimit] == null ? null : int.parse(headers[xRateLimitLimit]!);
 
-      final response = e.response as http.StreamedResponse;
+  static int? getRemainingFromHeaders(Map<String, String> headers) => headers[xRateLimitRemaining] == null ? null : int.parse(headers[xRateLimitRemaining]!);
 
-      // Check for 429, emmit events and wait given in response body time
-      if (response.statusCode == 429) {
-        final responseBody = jsonDecode(await response.stream.bytesToString());
-        final retryAfter = ((responseBody["retry_after"] as double) * 1000).round();
+  // Server-Client clock drift makes headers.reset useless, build reset from headers.resetAfter and DateTime.now()
+  static DateTime? getResetFromHeaders(Map<String, String> headers) => headers[xRateLimitResetAfter] == null ? null : DateTime.now().add(getResetAfterFromHeaders(headers)!);
 
-        _events.onRateLimitedController.add(RatelimitEvent(request, false, response));
-        _httpHandler.logger.warning("Rate limited via 429 on endpoint: ${request.uri}. Trying to send request again in $retryAfter ms...");
+  static Duration? getResetAfterFromHeaders(Map<String, String> headers) => headers[xRateLimitResetAfter] == null ? null : Duration(milliseconds: (double.parse(headers[xRateLimitResetAfter]!) * 1000).ceil());
 
-        return Future.delayed(Duration(milliseconds: retryAfter), () => execute(request));
-      }
+  void addInFlightRequest(HttpRequest httpRequest) => _inFlightRequests.add(httpRequest);
 
-      // Return http error
-      _setBucketValues(response.headers);
-      return response;
-    }
+  void removeInFlightRequest(HttpRequest httpRequest) => _inFlightRequests.remove(httpRequest);
+
+  bool isInBucket(http.StreamedResponse response) {
+    return getBucketIdFromHeaders(response.headers) == _bucketId;
   }
 
-  void _setBucketValues(Map<String, String> headers) {
-    if (headers["x-ratelimit-remaining"] != null) {
-      _remaining = int.parse(headers["x-ratelimit-remaining"]!);
-    }
+  updateRateLimit(http.StreamedResponse response) {
+    if (!isInBucket(response)) {
+      return;
```

Comment:
if there is return then you dont need else here. Would be more readable to flatten if  statements